// @elastic/eui/es/components/flyout/flyout_resizable.styles.js
function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { css } from '@emotion/react';
import { logicalCSS } from '../../global_styling';
var _ref = process.env.NODE_ENV === "production" ? {
  name: "1gfwqnd-euiFlyoutResizableButton",
  styles: "position:absolute;label:euiFlyoutResizableButton;"
} : {
  name: "1gfwqnd-euiFlyoutResizableButton",
  styles: "position:absolute;label:euiFlyoutResizableButton;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};
export var euiFlyoutResizableButtonStyles = function euiFlyoutResizableButtonStyles(_ref2) {
  var euiTheme = _ref2.euiTheme;
  return {
    euiFlyoutResizableButton: _ref,
    overlay: {
      left: /*#__PURE__*/css(logicalCSS('right', 0), ";;label:left;"),
      right: /*#__PURE__*/css(logicalCSS('left', 0), ";;label:right;")
    },
    push: {
      left: /*#__PURE__*/css(logicalCSS('right', "-".concat(euiTheme.border.width.thin)), ";;label:left;"),
      right: /*#__PURE__*/css(logicalCSS('left', "-".concat(euiTheme.border.width.thin)), ";;label:right;")
    }
  };
};
